### PR TITLE
feat: add Quine–McCluskey algorithm

### PR DESCRIPTION
**Pull request by jang-jaehyuk**
#### Checklist

- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1845"></a>

### CODE REVIEW

File: bit_manipulation/quine_macluskey.cpp [hunk 1]
```diff
@@ -0,0 +1,404 @@
+/**
+ * @file quine_macluskey.cpp
```
Comment (Panquesito7) [inline bit_manipulation/quine_macluskey.cpp:2]:
```suggestion
 * @file
```

File: bit_manipulation/quine_macluskey.cpp [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+ * @author [jang-jaehyuk](https://github.com/jang-jaehyuk)
+ * @brief Implementation of [minimization of
+ * Boolean functions](https://en.wikipedia.org/wiki/Quine-McCluskey_algorithm).
+ */
+
+#include <iostream> /// for io operations
+#include <vector> /// for store the implicant
+#include <algorithm> /// for sorting implicant
+
+using namespace std;
```
Comment (Panquesito7) [inline bit_manipulation/quine_macluskey.cpp:12]: Let use `using std::cout` instead. 🙂
```suggestion
using std::cout;
using std::vector;
using std::cin;
using std::endl;
```

If I missed any, please let me know.

File: bit_manipulation/quine_macluskey.cpp [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+			if ((i < epi.size() - 1) && (!equalArrays(epi[i].v, epi[i + 1].v, min(epi[i].v.size(), epi[i + 1].v.size())))) {
+				printf(" + ");
+			}
+			else if ((i < epi.size() - 1) && (equalArrays(epi[i].v, epi[i + 1].v, min(epi[i].v.size(), epi[i + 1].v.size())))) {
+				printf(" or ");
+			}
+		}
+	}
+};
+int main() {
```
Comment (Panquesito7) [inline bit_manipulation/quine_macluskey.cpp:374]:
```suggestion

/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: bit_manipulation/quine_macluskey.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+			}
+			if ((i < epi.size() - 1) && (!equalArrays(epi[i].v, epi[i + 1].v, min(epi[i].v.size(), epi[i + 1].v.size())))) {
+				printf(" + ");
+			}
+			else if ((i < epi.size() - 1) && (equalArrays(epi[i].v, epi[i + 1].v, min(epi[i].v.size(), epi[i + 1].v.size())))) {
+				printf(" or ");
+			}
+		}
+	}
+};
```
Comment (Panquesito7) [inline bit_manipulation/quine_macluskey.cpp:373]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or in our [**Discord server**](https://discord.gg/c7MnfGFGa6). 😃

File: bit_manipulation/quine_macluskey.cpp [hunk 5]
```diff
@@ -0,0 +1,404 @@
+/**
+ * @file quine_macluskey.cpp
+ * @author [jang-jaehyuk](https://github.com/jang-jaehyuk)
+ * @brief Implementation of [minimization of
+ * Boolean functions](https://en.wikipedia.org/wiki/Quine-McCluskey_algorithm).
```
Comment (Panquesito7) [inline bit_manipulation/quine_macluskey.cpp:5]: Please add a detailed description of how the algorithm works, what does it do, etc..
Let us know if you need any help here or in our [**Discord server**](https://discord.gg/c7MnfGFGa6). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/jang-jaehyuk/C-Plus-Plus/actions

File: .github/workflows/cmake.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+name: CMake
```
Comment (Panquesito7) [inline .github/workflows/cmake.yml:1]: Please remove this file. We already have a workflow that runs CMake commands. https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/.github/workflows/awesome_workflow.yml

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: .github/workflows/cmake.yml (hunk 6 above)
Comment (jang-jaehyuk) [reply .github/workflows/cmake.yml:1]: Disabled Awesome CI Workflow in action in my repository!
did you say this?

File: bit_manipulation/quine_mccluskey.cpp [hunk 7]
```diff
@@ -0,0 +1,459 @@
+/**
+ * @file quine_mccluskey.cpp
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:2]:
```suggestion
 * @file
```

File: bit_manipulation/quine_mccluskey.cpp [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+ * @author [jang-jaehyuk](https://github.com/jang-jaehyuk)
+ * @brief Implementation of [minimization of
+ * Boolean functions](https://en.wikipedia.org/wiki/Quine-McCluskey_algorithm)
+ * @details The Quine-McCluskey algorithm is an algorithm that minimizes logical expressions.
+ * The minimum form of a logical function can be obtained deterministically.
+ * f_{A,B,C,D}=BC'D'+AB'+AC
+ * The above expression minimizes the following expression and is logically equivalent.
+ * f_{A,B,C,D}=A'BC'D'+AB'C'D'+AB'C'D+AB'CD'+AB'CD+ABC'D'+ABCD'+ABCD
+ * The algorithm goes through two main steps.
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:11]:
```suggestion
 *
 * ### Algorithm
```

File: bit_manipulation/quine_mccluskey.cpp [hunk 9]
```diff
@@ -0,0 +1,459 @@
+/**
+ * @file quine_mccluskey.cpp
+ * @author [jang-jaehyuk](https://github.com/jang-jaehyuk)
+ * @brief Implementation of [minimization of
+ * Boolean functions](https://en.wikipedia.org/wiki/Quine-McCluskey_algorithm)
+ * @details The Quine-McCluskey algorithm is an algorithm that minimizes logical expressions.
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:6]:
```suggestion
 * Boolean functions](https://en.wikipedia.org/wiki/Quine-McCluskey_algorithm)
 *
 * @details
 * The Quine-McCluskey algorithm is an algorithm that minimizes logical expressions.
```

File: bit_manipulation/quine_mccluskey.cpp [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+ * The algorithm goes through two main steps.
+ * 1. Find all the Prime Implicants of a given function.
+ * 2. Find the Essential Prime Implicant from the candidate term table using the candidate terms.
+ */
+
+#include <iostream> /// for io operations
+#include <vector> /// for store the implicant
+#include <algorithm> /// for sorting implicant
+#include <string>	/// for string
+#include <cassert>   /// for assert
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:20]:
```suggestion
#include <iostream>  /// for IO operations
#include <vector>    /// for std::vector
#include <algorithm> /// for sorting implicant
#include <string>    /// for string operations
#include <cassert>   /// for assert
```

File: bit_manipulation/quine_mccluskey.cpp [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+	int ex3[] = { 1, 2, 3, 4, 5, 7, 9, 15 };
+	cnt = sizeof(ex3) / sizeof(int);
+	assert(out3.compare(quine.quineExe(ex3, cnt)));
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+	test();
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:457]:
```suggestion
	test();  // run self-test implementations
```

File: bit_manipulation/quine_mccluskey.cpp [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+
+	McCluskey quine;
+	int cnt = 0;
+	string out1("B'C' + CD' + A'BD");	///expected output
+	string out2("B'C' + BC + A'B' or A'C"); ///expected output
+	string out3("B'C'D + A'B'C + A'BC' + BCD"); ///expected output
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:437]:
```suggestion
	string out1("B'C' + CD' + A'BD");	/// expected output
	string out2("B'C' + BC + A'B' or A'C"); /// expected output
	string out3("B'C'D + A'B'C + A'BC' + BCD"); /// expected output
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by jang-jaehyuk** [comment]
Another problem? I am very much looking forward to my first contribution!

File: bit_manipulation/quine_mccluskey.cpp (hunk 12 above)
Comment (jang-jaehyuk) [reply bit_manipulation/quine_mccluskey.cpp]: all corrected!

please check.

File: bit_manipulation/quine_mccluskey.cpp [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+using std::cin;
+using std::endl;
+using std::min;
+using std::string;
+
+/**
+ * @brief This structure is for organizing the bits.
+ */
+typedef struct BitNode {
+	vector<int> v; /// combined numbers or selected number
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:35]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6)!** 😃

File: bit_manipulation/quine_mccluskey.cpp [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+	int a = 0;	/// first bit value 
+	int b = 0;	/// second bit value 
+	int c = 0;	/// third bit value 
+	int d = 0;	/// fourth bit value 
+	int howmany1 = 0;	/// number of 1
```
Comment (Panquesito7) [inline bit_manipulation/quine_mccluskey.cpp:40]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6)!** 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4487816644?check_suite_focus=true) warnings. Let us know if you need any help. 😃

**Pull request closed without merging**